### PR TITLE
Feature/blockchain tests

### PR DESCRIPTION
**Pull request by davikstone2**
Adds a few tests relating to token transfer and token persistence

### CODE REVIEW

File: saito-core/Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "saito-core"
-version = "0.1.3"
+version = "0.1.4"
```
Comment (SankaD) [inline saito-core/Cargo.toml:3]: no need to update version if it's only test related changes

**Comment by SankaD** [comment]
Please fix the compile errors.

**Comment by davikstone2** [comment]
> Please fix the compile errors.

fixed and updated.

File: saito-core/src/common/defs.rs [hunk 2]
```diff
@@ -19,6 +19,8 @@ pub type BlockId = u64;
 
 pub const NOLAN_PER_SAITO: Currency = 100_000_000;
 
+pub const ISSUANCE_PUBLIC_KEY: &'static str = "q6TTBeSStCLXEPoS5TUVAxNiGGnRDZQenpvAXXAfTmtA";
```
Comment (SankaD) [inline saito-core/src/common/defs.rs:22]: what is this key? why it's hard coded ?

File: saito-core/src/common/test_manager.rs [hunk 3]
```diff
@@ -48,9 +51,10 @@ pub mod test {
     use crate::core::data::mempool::Mempool;
     use crate::core::data::network::Network;
     use crate::core::data::peer_collection::PeerCollection;
+    use crate::core::data::slip::{Slip, SlipType};
     use crate::core::data::storage::Storage;
     use crate::core::data::transaction::{Transaction, TransactionType};
-    use crate::core::data::wallet::Wallet;
+    use crate::core::data::wallet::{self, Wallet};
```
Comment (SankaD) [inline saito-core/src/common/test_manager.rs:57]: why is self here?

File: saito-core/src/common/defs.rs (hunk 2 above)
Comment (SankaD) [reply saito-core/src/common/defs.rs:22]: rename to project public key perhaps?

File: saito-core/src/common/test_manager.rs [hunk 4]
```diff
@@ -537,17 +572,157 @@ pub mod test {
...
+
+        pub fn get_wallet_lock(&self) -> Arc<RwLock<Wallet>> {
+            return self.wallet_lock.clone();
+        }
+
+        pub async fn get_wallet(&self) -> tokio::sync::RwLockReadGuard<'_, Wallet> {
+            let wallet;
+            let _wallet_;
+            (wallet, _wallet_) = lock_for_read!(self.wallet_lock, LOCK_ORDER_WALLET);
+            return wallet;
```
Comment (SankaD) [inline saito-core/src/common/test_manager.rs:587]: this shouldn't be returned like this. until returned wallet is destroyed lock is kept.

File: saito-core/src/core/data/blockchain.rs [hunk 5]
```diff
@@ -1643,6 +1643,14 @@ impl Blockchain {
 
         snapshot
     }
+    pub fn add_to_utxo(&mut self, slip: &Slip) {
+        self.utxoset.insert(slip.get_utxoset_key(), true);
+    }
+    pub fn update_from_balance_snapshot(&mut self, snapshot: BalanceSnapshot) {
```
Comment (SankaD) [inline saito-core/src/core/data/blockchain.rs:1649]: rebase from develop please

File: saito-core/src/core/data/blockchain.rs [hunk 6]
```diff
@@ -1664,7 +1672,7 @@ mod tests {
     use crate::core::data::slip::Slip;
     use crate::core::data::wallet::Wallet;
     use crate::{lock_for_read, lock_for_write};
-
+    const ISSUANCE_FILE_PATH: &'static str = "./data/issuance/test/issuance";
```
Comment (SankaD) [inline saito-core/src/core/data/blockchain.rs:1675]: rename to test issuance file path ? and define in test manager?

File: saito-core/src/core/data/issuance [hunk 7]
```diff
@@ -0,0 +1,4 @@
+5810123	n5hNA7Tazy7D226nq5arWq5YHxe8ibXGqNtjwXpEhnPT	VipOutput
```
Comment (SankaD) [inline saito-core/src/core/data/issuance:1]: there are too many issuance files now. please move them to a single location

File: saito-core/src/core/data/storage.rs [hunk 8]
```diff
@@ -240,7 +248,7 @@ impl Storage {
         }
     }
 
-    fn decode_str(string: &str) -> Result<Vec<u8>, bs58::decode::Error> {
+    pub fn decode_str(&self, string: &str) -> Result<Vec<u8>, bs58::decode::Error> {
```
Comment (SankaD) [inline saito-core/src/core/data/storage.rs:251]: don't need &self for this

File: saito-core/src/core/data/storage.rs [hunk 9]
```diff
@@ -281,29 +289,57 @@ impl Storage {
...
 
     use crate::common::defs::{SaitoHash, MAX_TOKEN_SUPPLY};
     use crate::common::test_manager::test::{create_timestamp, TestManager};
     use crate::core::data::block::Block;
     use crate::core::data::crypto::{hash, verify};
+    use crate::core::data::storage::Storage;
+    // use std::io::{BufRead, BufReader};
 
-    #[ignore]
+    const ISSUANCE_FILE_PATH: &'static str = "./data/issuance/test/issuance";
```
Comment (SankaD) [inline saito-core/src/core/data/storage.rs:301]: define in test_manager or some common place

File: saito-core/src/core/data/transaction.rs [hunk 10]
```diff
@@ -155,7 +155,7 @@ impl Transaction {
         mut with_fee: Currency,
         _force_merge: bool,
     ) -> Result<Transaction, Error> {
-        trace!(
+        dbg!(
```
Comment (SankaD) [inline saito-core/src/core/data/transaction.rs:158]: debug!()

**Review by SankaD — CHANGES_REQUESTED** [review]

File: saito-core/src/core/data/issuance (hunk 7 above)
Comment (davik20) [reply saito-core/src/core/data/issuance:1]: The other issuance file is for testing purposes

File: saito-core/src/common/defs.rs [hunk 11]
```diff
@@ -19,9 +19,11 @@ pub type BlockId = u64;
 
 pub const NOLAN_PER_SAITO: Currency = 100_000_000;
 
+pub const PROJECT_PUBLIC_KEY: &'static str = "q6TTBeSStCLXEPoS5TUVAxNiGGnRDZQenpvAXXAfTmtA";
+
 // length of 1 genesis period
 pub const GENESIS_PERIOD: u64 = 100_000;
-// prune blocks from index after N blocks
+// prune blocks from index after TN blocks
```
Comment (SankaD) [inline saito-core/src/common/defs.rs:26]: T?

File: saito-core/src/core/data/issuance (hunk 7 above)
Comment (SankaD) [reply saito-core/src/core/data/issuance:1]: there's a saito-core/src/common/issuance also

File: saito-core/src/core/data/transaction.rs [hunk 12]
```diff
@@ -168,16 +167,17 @@ impl Transaction {
         }
 
         let total_requested = with_payment + with_fee;
-        trace!(
+        dbg!(
```
Comment (SankaD) [inline saito-core/src/core/data/transaction.rs:170]: debug!

File: saito-core/src/core/data/transaction.rs [hunk 13]
```diff
@@ -168,16 +167,17 @@ impl Transaction {
...
             "in generate transaction. available: {} and payment: {} and fee: {}",
             available_balance,
             with_payment,
             with_fee
         );
         if available_balance < total_requested {
-            warn!(
+            dbg!(
```
Comment (SankaD) [inline saito-core/src/core/data/transaction.rs:177]: debug!

File: saito-core/src/core/data/transaction.rs [hunk 14]
```diff
@@ -1067,7 +1067,7 @@ mod tests {
         assert_eq!(tx.timestamp, 0);
         assert_eq!(tx.from, vec![]);
         assert_eq!(tx.to, vec![]);
-        assert_eq!(tx.data, vec![]);
+        // assert_eq!(tx.data, vec![]);
```
Comment (SankaD) [inline saito-core/src/core/data/transaction.rs:1070]: why commented?

**Review by SankaD — CHANGES_REQUESTED** [review]

**Comment by arpee** [comment]
@SankaD and @davik20 - can you please let me now when this is completed so that I can test.

Thanks

**Review by SankaD — APPROVED** [review]

**Pull request merged**